### PR TITLE
Feature/hpj/permission

### PR DESCRIPTION
**Pull request by flybikeGx**

### CODE REVIEW

File: vm/host/authority.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	u := p.Users
+	for _, g := range p.Groups {
+		grp, ok := a.Groups[g]
+		if !ok {
+			continue
+		}
+		u = append(u, grp.Users...)
+	}
```
Comment (wangyu0air) [inline vm/host/authority.go]: what if user duplicated

File: vm/host/authority.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	for _, user := range u {
+		if user.IsKeyPair {
+			if _, ok := auth[user.ID]; ok {
+				weight += user.Weight
+				if weight >= p.Threshold {
+					return true, c
+				}
+			}
+		} else {
+			ok, cost := h.requireAuth(user.ID, user.Permission, auth)
```
Comment (wangyu0air) [inline vm/host/authority.go]: check requireAuth reentrance

Comment (flybikeGx) [reply vm/host/authority.go]: fixed

File: vm/host/authority.go (hunk 1 above)
Comment (flybikeGx) [reply vm/host/authority.go]: fixed

File: contract/account.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                id : un.substring(0, len),
+                permission: un.substring(len, un.length()),
+                is_key_pair: false,
+                weight: weight
+            });
+        }
+
+        this._saveAccount(acc)
+    }
+    AddPermissionToGroup(id, perm, group) {
```
[the remark this answers is not in the document]
Comment (flybikeGx) [reply contract/account.js]: fixed

File: contract/account.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            });
+        } else {
+            acc.permissions[perm].users.push({
+                id : un.substring(0, len),
+                permission: un.substring(len, un.length()),
+                is_key_pair: false,
+                weight: weight
+            });
+        }
+
```
[the remark this answers is not in the document]
Comment (flybikeGx) [reply contract/account.js]: fixed

File: contract/account.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            users: [{
+                id: owner,
+                is_key_pair: true,
+                weight: 1,
+            }],
+            threshold: 1,
+        };
+        this._saveAccount(account)
+    }
+    AddPermission(id, perm, thres) {
```
[the remark this answers is not in the document]
Comment (flybikeGx) [reply contract/account.js]: fixed

**Pull request merged**